proxy_agent/src/proxy/proxy_server.rs

```diff
@@ -0,0 +1,694 @@
+use proxy_agent_shared::proxy_agent_aggregate_status::ProxyAgentDetailStatus;
+use proxy_agent_shared::telemetry::event_logger;
+use std::sync::{Arc, Mutex};
+use tokio::net::TcpListener;
+use tokio::net::TcpStream;
+
+const INITIAL_CONNECTION_ID: u128 = 0;
+
+pub fn stop(port: u16, shared_state: Arc<Mutex<SharedState>>) {
+    proxy_listener_wrapper::set_shutdown(shared_state.clone(), true);
+    let _ = std::net::TcpStream::connect(format!("127.0.0.1:{}", port));
+    logger::write_warning("Sending stop signal.".to_string());
+}
+
+pub fn get_status(shared_state: Arc<Mutex<SharedState>>) -> ProxyAgentDetailStatus {
+    let status = if proxy_listener_wrapper::get_shutdown(shared_state.clone()) {
+        ModuleState::STOPPED.to_string()
+    } else {
+        ModuleState::RUNNING.to_string()
+    };
+
+    ProxyAgentDetailStatus {
+        status,
+        message: proxy_listener_wrapper::get_status_message(shared_state.clone()),
+        states: None,
+    }
+}
+
+pub async fn start_async(port: u16, shared_state: Arc<Mutex<SharedState>>) {
+    tokio::spawn(async move { start(port, shared_state).await });
+}
+
+async fn start(
+    port: u16,
+    shared_state: Arc<Mutex<SharedState>>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    Connection::init_logger(config::get_logs_dir());
+
+    let addr = format!("{}:{}", std::net::Ipv4Addr::LOCALHOST, port);
+    logger::write(format!("Start proxy listener at '{}'.", &addr));
+
+    let listener = match TcpListener::bind(&addr).await {
+        Ok(listener) => listener,
+        Err(e) => {
+            let message = format!("Failed to bind TcpListener '{}' with error {}.", addr, e);
+            proxy_listener_wrapper::set_status_message(shared_state.clone(), message.to_string());
+            logger::write_error(message);
+            return Err(Box::new(e));
+        }
+    };
+
+    let message = helpers::write_startup_event(
+        "Started proxy listener, ready to accept request",
+        "start",
+        "proxy_listener",
+        logger::AGENT_LOGGER_KEY,
+    );
+    proxy_listener_wrapper::set_status_message(shared_state.clone(), message.to_string());
+    provision::listener_started(shared_state.clone()).await;
+
+    // We start a loop to continuously accept incoming connections
+    loop {
+        let (stream, client_addr) = match listener.accept().await {
+            Ok((stream, client_addr)) => (stream, client_addr),
+            Err(e) => {
+                logger::write_warning(format!("ProxyListener accept error {}", e));
+                continue;
+            }
+        };
+
+        if shared_state_wrapper::get_cancellation_token(shared_state.clone()).is_cancelled() {
+            let message = "Stop signal received, stop the listener.";
+            proxy_listener_wrapper::set_status_message(shared_state.clone(), message.to_string());
+            logger::write_warning(message.to_string());
+            return Ok(());
+        }
```

Comment:
This is not really how the cancellation token is meant to be used. Instead, you can `select!` on the cancellation token along with the future you want to let someone else cancel.

Reply:
I switched to Shutdown state in shared_state as before.